### PR TITLE
Batch heartbeats from node manager together in the monitor.

### PR DESCRIPTION
**Pull request by ujvl**
Instead of sending n^2 heartbeats from each node manager to each other node manager, send 2n heartbeats by sending heartbeats from each node manager to the monitor, batching them together in the monitor, and sending them out to all the node managers. #2949

### CODE REVIEW

File: src/ray/gcs/client.h
Comment (robertnishihara) [inline src/ray/gcs/client.h]: You seem to have added a bunch of new lines like this. I'm ok with it either way assuming it passes linting, but is there a reason for this?

File: src/ray/raylet/monitor.cc
Comment (robertnishihara) [inline src/ray/raylet/monitor.cc]: This does a copy? Just want to make sure.

**Comment by robertnishihara** [comment]
Note I updated the PR description.

File: src/ray/raylet/node_manager.cc
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc]: Looking at the code, it doesn't seem like `client` is actually used in either this method or `HeartbeatAdded`. Can you remove the argument from both methods?

Comment (robertnishihara) [reply src/ray/raylet/node_manager.cc]: Neither is `publish_key`. Should we remove that as well? I'm also confused about what that actually is.

File: src/ray/raylet/node_manager.h
Comment (robertnishihara) [inline src/ray/raylet/node_manager.h]: Document the args

File: src/ray/raylet/node_manager.cc
Comment (ujvl) [reply src/ray/raylet/node_manager.cc]: Oh hm yeah I just cherry-picked our commits from the benchmarks, I'll take a pass through.

**Comment by robertnishihara** [comment]
@ujvl what is the status of this PR? Is it waiting for review or are there more things to do?

**Comment by ujvl** [comment]
@robertnishihara one test (`test_actor_multiple_gpus_from_multiple_tasks`) is failing, looking into that.

File: test/actor_test.py
Comment (robertnishihara) [inline test/actor_test.py]: @stephanie-wang @ujvl I just looked at this test, and the doesn't make any sense. It really shouldn't have been passing before (except it probably passed because the timeout in the line `ready_ids, _ = ray.wait([a.get_location_and_ids.remote()], timeout=10)` was so short.

But when a `create_actors` task finishes, it's actors should go out of scope and so all the GPUs will get freed up again.

I'm not sure if this test is testing anything useful. Maybe we should just remove it?

Comment (ujvl) [reply test/actor_test.py]: Hmm yeah I'm not sure if the test is written correctly, but currently I'm trying to debug performance; the test seems to hang for a long time **inside** `create_actors` (similar to #3088 ?) before it even gets to the asserts.

Comment (robertnishihara) [reply test/actor_test.py]: Interesting. It's trying to create 100 actors, so there's going to be some rate-limiting based on the number of CPUs.

If it's related to this PR, then we should fix it of course, but if not, then maybe we should get this merged and look into it separately?

Comment (ujvl) [reply test/actor_test.py]: Yeah I'm just unsure if this PR is exacerbating some existing problem or creating a new one, the test passes on `master` but not on this branch.

On master the test finishes immediately whereas on this branch there are > `50s` gaps between actor creations, even if I reduce the number of actors.

Comment (stephanie-wang) [reply test/actor_test.py]: Hmm yeah, it might make sense to separate out the fix for the unit test into its own PR. I think the PR was passing on `master`, but I think it was also significantly slower than without xray. Can't remember for sure right now.

File: src/ray/raylet/node_manager.cc
Comment (robertnishihara) [reply src/ray/raylet/node_manager.cc]: or at least const

**Comment by stephanie-wang** [comment]
@ujvl, what is the status of this PR? Did you try out reordering the order of heartbeats in the batch to match the order that the monitor received them in?

**Comment by atumanov** [comment]
```
$ while [ 1 ]; do python -m pytest test/actor_test.py::test_actor_multiple_gpus_from_multiple_tasks ; done
======================================================================= test session starts =======================================================================
platform darwin -- Python 2.7.12, pytest-3.5.0, py-1.5.3, pluggy-0.6.0
rootdir: /Users/atumanov/devrepo/github/ray-mydev, inifile:
collected 1 item

test/actor_test.py .                                                                                                                                        [100%]

==================================================================== 1 passed in 38.26 seconds ====================================================================
```

Running this continuously in a loop seems to be passing all the time on my macbook.

**Review by atumanov — COMMENTED** [review]
The PR looks good to me, but I think we should adjust the spillover policy, either as part of this PR or a separate PR, to distribute infeasible tasks across all feasible candidates in the heartbeat batch.

**Comment by stephanie-wang** [comment]
The test is failing on [travis](https://travis-ci.com/ray-project/ray/jobs/154147283).

**Comment by ujvl** [comment]
Yeah spillover should be adjusted in a separate PR; still need to figure out why it fails on Travis on this one though. Also even if it's passing there might still be a high task latency (although ordering the heartbeats by arrival should have fixed that).

File: src/ray/raylet/scheduling_policy.h
Comment (stephanie-wang) [inline src/ray/raylet/scheduling_policy.h]: Is the change to consider the entire batch in `SpillOver` necessary? I prefer if we don't modify the scheduling policy at all and instead randomize the order of the batch in `HeartbeatBatchAdded`.

File: src/ray/raylet/scheduling_policy.cc
Comment (stephanie-wang) [inline src/ray/raylet/scheduling_policy.cc]: Can the shuffle be done in the monitor, or does it need to be performed independently at every node?

Comment (ujvl) [reply src/ray/raylet/scheduling_policy.cc]: It can be done in the monitor (which I tested and found to work) but my reasoning was that it doesn't make sense to shuffle there since the monitor isn't responsible for fixing a scheduling policy issue. But yeah if I move it there it does work.

Comment (stephanie-wang) [reply src/ray/raylet/scheduling_policy.cc]: I think we should probably put the shuffle in the monitor for now and avoid changes to the scheduling policy in this PR.

File: src/ray/raylet/scheduling_policy.h
Comment (ujvl) [reply src/ray/raylet/scheduling_policy.h]: I personally think the Scheduling policy should take advantage of having the entire batch to make better decisions; even though we're shuffling the vector here there could be a better solution, in which case it's better to leave it to the policy to make the decision of whether or not randomizing order is necessary. 

Randomizing it outside is possible, but to your other comment in that case we might as well just randomize at the monitor and save the nodes from the work.

Comment (stephanie-wang) [reply src/ray/raylet/scheduling_policy.h]: I agree, but I think we should leave it for a future PR and just try to emulate the functionality on master right now. I'm actually not so concerned about whether the randomization is expensive, just want to keep the PR as simple as possible.

File: src/ray/raylet/scheduling_policy.cc
Comment (atumanov) [inline src/ray/raylet/scheduling_policy.cc]: The random shuffle of remote client ids above doesn't address the problem. In this for loop you are iterating over the remote client ids in the same order for all infeasible tasks. The result of this is still sending all infeasible tasks to the first remote client id in the shuffled list.

File: src/ray/raylet/node_manager.cc
Comment (atumanov) [inline src/ray/raylet/node_manager.cc]: OK, this is where the problem is introduced. You don't want to spillover to the entire cluster. You ONLY want to spillover to the heartbeating raylets. Similarly to how the cluster map was constructed in the original HeartbeatAdded code, you need to construct a cluster map for a heartbeating batch of nodes and pass _that_ to SpillOver, not the whole cluster resource map.

**Review by atumanov — DISMISSED** [review]
OK, this PR needs the following changes:
1. construct a resource map for heartbeating batch only. Do not pass the whole cluster resource map to spillover too -- this introduced the pathological circular behavior I think you are seeing
2. Inside the Spillover policy shuffle is not enough. Need to load balance infeasible tasks over all feasible candidates in the batch.

I can help with implementing these changes.

File: src/ray/raylet/scheduling_policy.cc
Comment (ujvl) [reply src/ray/raylet/scheduling_policy.cc]: Yeah you're right, I initially toyed with generating indices uniformly at random to distribute the load more uniformly, but it was getting too complicated: for example, if you try to spill over to a random node but it doesn't have the set of resources required, we'd have to try again (or give up) on it etc. 

Figured that shuffling would be a quick fix to enforce that between heartbeats things are random so any given task is less likely to be forwarded in cycles. But yes, over a single heartbeat it's going to end up dumping everything on one node.

File: src/ray/raylet/node_manager.cc
Comment (ujvl) [reply src/ray/raylet/node_manager.cc]: I passed in the cluster map just so I didn't have to deal with copying KV pairs to a new map. In the function I only use the keys corresponding to `client_ids` which does what you're describing I think.

**Review by stephanie-wang — APPROVED** [review]
Pushed some linting changes, but the travis tests looked good. I'll merge once the linting commit passes.

**Pull request merged**
